### PR TITLE
Do not return a buffer when slurping a empty file

### PR DESCRIPTION
**Pull request by mscherer**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [X] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

While running r2 from git, I was greeted by:
    
    |ERROR| Invalid command ��Y�' (0x08)
    
I quickly found this came from:
    
    $ ls -l ~/.radare2rc
    -rw-r--r--. 1 misc misc 0 19 avril  2017 /home/misc/.radare2rc
    
Debugging showed that if the file is empty, we return the buffer
that was reallocate earlier with uninitialized memory.
    
Since r_core_cmd_file do not check the size (it pass NULL as a
2nd parameter for r_file_slurp), it doesn't see the file is empty,
and so try to interpret 6 bytes as a command.
    
Given that grep find 54 calls to r_file_slurp with NULL
as a 2nd parameter, this will also fix a bit more than just this
error message.

**Test plan**

* Create a empty file ~/.radare2rc
* Run r2
* See no error message about "|ERROR| Invalid command"

### CODE REVIEW

**Review by ret2libc — CHANGES_REQUESTED** [review]
I think this fix is not good from an API point of view. If I call `r_file_slurp` on an empty file I would expect an empty buffer to be returned, not NULL (which is usually used to indicate there was an error). I think the issue here is that `buf` is not NULL terminated, so random stuff is read from there.

I believe the correct fix is to ensure that buf gets NULL terminated even in the `if (!sz) {` path. So callers that do not pass/use `usz` can still work correctly (assuming the file is not a binary file and does not contain \x00 bytes).

**Comment by ret2libc** [comment]
> Ok, I will fix that. I will also add a unit test for that, as soon as I find the right place for them (it seems that ./test/unit/legacy_unit/ is a good place, but "legacy" do not look good, and didn't looked at the documentation yet for how to do it).

Please add the test in `test/unit/test_util.c`. Thanks!

File: libr/util/file.c [hunk 1]
```diff
@@ -365,6 +365,9 @@ R_API char *r_file_slurp(const char *str, R_NULLABLE size_t *usz) {
 				}
 				size += r;
 			} while (!feof (fd));
+			if (!size) {
```
Comment (ret2libc) [inline libr/util/file.c:368]: probably there should always be `buf[size] = '\0';`, no matter what's the value of size.

File: test/unit/test_util.c [hunk 2]
```diff
@@ -113,6 +113,34 @@ bool test_autonames(void) {
...
+	int f = open(nonempty_file, O_CREAT|O_WRONLY);
+	write(f, some_words);
+	close(f);
+
+	content = r_file_slurp(nonempty_file, &s);
+
+	mu_assert_eq (s, strlen(some_words), "size should be correct");
+	mu_assert_eq (strlen(content), strlen(some_words), "size for the buffer should be correct");
+	mu_assert_streq (s, some_words, "content should match");
+}
```
Comment (ret2libc) [inline test/unit/test_util.c:142]: add `mu_end` before final `}`

Comment (mscherer) [reply test/unit/test_util.c:145]: Done

**Review by ret2libc — APPROVED** [review]

File: test/unit/test_util.c [hunk 3]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
...
+#ifdef __WINDOWS__
+#define S_IRWXU _S_IREAD | _S_IWRITE
+#endif
+
+	char* test_file =  "./empty_file";
+	size_t s;
+	char* content;
+	char* some_words = "some words";
+
+	close(open(test_file, O_CREAT, S_IRWXU));
```
Comment (trufae) [inline test/unit/test_util.c:127]: missing space before (. also may be worth checking the return value of open to avoid calling close(-1) with an assert

File: test/unit/test_util.c [hunk 4]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
...
+
+	content = r_file_slurp(test_file, &s);
```
Comment (trufae) [inline test/unit/test_util.c:129]: defnie and assign the variable in the same line and add a space bfore (

File: test/unit/test_util.c [hunk 5]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
...
+
+	mu_assert_eq (s, 0, "size should be zero");
+	mu_assert_eq (strlen(content), 0, "returned buffer should be empty");
+
+	free(content);
+
+	int f = open(test_file, O_CREAT|O_WRONLY, S_IRWXU);
```
Comment (trufae) [inline test/unit/test_util.c:136]: check for the return value of open

Comment (trufae) [reply test/unit/test_util.c:136]: the file wasnt removed, so o_creat shuoldnt be correct tob e used here

File: libr/util/file.c [hunk 6]
```diff
@@ -365,6 +365,7 @@ R_API char *r_file_slurp(const char *str, R_NULLABLE size_t *usz) {
 				}
 				size += r;
 			} while (!feof (fd));
+			buf[size] = '\0';
```
Comment (trufae) [inline libr/util/file.c:368]: change the code of the function to just do a single malloc(sz+1) if you want to null terminate the buffer. the current implementation do realloc for no reason.

File: test/unit/test_util.c [hunk 7]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
...
+bool test_file_slurp(void) {
+
+#ifdef __WINDOWS__
+#define S_IRWXU _S_IREAD | _S_IWRITE
+#endif
+
+	char* test_file =  "./empty_file";
+	size_t s;
+	char* content;
+	char* some_words = "some words";
```
Comment (trufae) [inline test/unit/test_util.c:125]: this is const

File: test/unit/test_util.c [hunk 8]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
 	mu_end;
 }
 
+bool test_file_slurp(void) {
+
+#ifdef __WINDOWS__
+#define S_IRWXU _S_IREAD | _S_IWRITE
+#endif
+
+	char* test_file =  "./empty_file";
```
Comment (trufae) [inline test/unit/test_util.c:122]: this is also const

File: test/unit/test_util.c [hunk 9]
```diff
@@ -113,6 +113,42 @@ bool test_autonames(void) {
...
+	size_t s;
+	char* content;
+	char* some_words = "some words";
+
+	close(open(test_file, O_CREAT, S_IRWXU));
+
+	content = r_file_slurp(test_file, &s);
+
+	mu_assert_eq (s, 0, "size should be zero");
+	mu_assert_eq (strlen(content), 0, "returned buffer should be empty");
```
Comment (trufae) [inline test/unit/test_util.c:132]: well. assuming the file is zero, the buffer shuold not contain anything, not even a null byte, because files are binary data not text. But i can get this assumption could be "handy" for some usages of the api

**Review by trufae — CHANGES_REQUESTED** [review]

File: test/unit/test_util.c (hunk 9 above)
Comment (ret2libc) [reply test/unit/test_util.c:132]: There are many cases where `r_file_slurp` is used assuming that it returns a null-terminated string. This is why I suggested this behaviour (which by the way already always holds except when `sz == 0`)

File: test/unit/test_util.c (hunk 7 above)
Comment (mscherer) [reply test/unit/test_util.c]: done

File: test/unit/test_util.c (hunk 8 above)
Comment (mscherer) [reply test/unit/test_util.c]: also done

File: test/unit/test_util.c (hunk 5 above)
Comment (mscherer) [reply test/unit/test_util.c]: done (both)

File: test/unit/test_util.c (hunk 3 above)
Comment (mscherer) [reply test/unit/test_util.c]: done, even if I am unsure on how we want to react exactly if open fail. I used some assert on the value after a quick reading of others tests cases, but I will dig further.

File: test/unit/test_util.c (hunk 4 above)
Comment (mscherer) [reply test/unit/test_util.c]: done

File: libr/util/file.c [hunk 10]
```diff
@@ -365,6 +365,11 @@ R_API char *r_file_slurp(const char *str, R_NULLABLE size_t *usz) {
 				}
 				size += r;
 			} while (!feof (fd));
+			char *nbuf = realloc (buf, size + 1);
+			if (nbuf) {
```
Comment (ret2libc) [inline libr/util/file.c:369]: I think you should invert the condition and return Null when the reallocation fails.

File: test/unit/test_util.c [hunk 11]
```diff
@@ -113,6 +113,41 @@ bool test_autonames(void) {
...
+    mu_assert_neq (f, -1, "cannot create empty file");
+    close (f);
+
+	char* content = r_file_slurp (test_file, &s);
+	mu_assert_eq (s, 0, "size should be zero");
+	mu_assert_eq (strlen(content), 0, "returned buffer should be empty");
+	free (content);
+
+	f = open (test_file, O_WRONLY, S_IRWXU);
+    mu_assert_neq (f, -1, "cannot reopen empty file");
```
Comment (ret2libc) [inline test/unit/test_util.c:136]: Pay attention to style please

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: test/unit/test_util.c (hunk 11 above)
Comment (mscherer) [reply test/unit/test_util.c:136]: Oups, I keep forgetting due to my vim setting :/

File: libr/util/file.c (hunk 6 above)
Comment (ret2libc) [reply libr/util/file.c:374]: I believe it does `realloc` because the size is not known. If it was known, we would not enter this path.

**Review by ret2libc — APPROVED** [review]

File: libr/util/file.c (hunk 6 above)
Comment (trufae) [reply libr/util/file.c:374]: the size should be known if its a regular file, only edge case are the files in /proc i think. i didnt tried the code, but it was a blind guess and its unrelated to the PR

File: test/unit/test_util.c [hunk 12]
```diff
@@ -113,6 +113,41 @@ bool test_autonames(void) {
...
+	mu_assert_eq (strlen(content), 0, "returned buffer should be empty");
+	free (content);
+
+	f = open (test_file, O_WRONLY, S_IRWXU);
+	mu_assert_neq (f, -1, "cannot reopen empty file");
+	write (f, some_words, strlen(some_words));
+	close (f);
+
+	content = r_file_slurp(test_file, &s);
+	mu_assert_eq (s, strlen(some_words), "size should be correct");
```
Comment (trufae) [inline test/unit/test_util.c:141]: some missing spaces before ( in this file

File: libr/util/file.c (hunk 6 above)
Comment (ret2libc) [reply libr/util/file.c:374]: yes indeed the general case is not this path. In general it just does `malloc(sz + 1)`, but this function tries to deal also with some files where it can't determine the size ahead of time, so continuous reallocation are needed.

**Review by ret2libc — APPROVED** [review]
@trufae please review again. It looks good to me.

File: libr/util/file.c [hunk 13]
```diff
@@ -365,6 +365,13 @@ R_API char *r_file_slurp(const char *str, R_NULLABLE size_t *usz) {
 				}
 				size += r;
 			} while (!feof (fd));
+			char *nbuf = realloc (buf, size + 1);
+			if (!nbuf) {
+				free(buf);
```
Comment (radare) [inline libr/util/file.c:370]: missing space before (

**Review by radare — APPROVED** [review]
Fix the indent and lgtm

File: libr/util/file.c (hunk 13 above)
Comment (ret2libc) [inline libr/util/file.c:370]:
```suggestion
				free (buf);
```

**Comment by ret2libc** [comment]
I did fix the space. Let's wait for CI and then let's merge.

**Comment by ret2libc** [comment]
Thanks!!
